src/extensions/surveys.tsx

```diff
@@ -444,6 +445,51 @@ export function usePopupVisibility(
         }
     }, [])
 
+    // Add URL change listener to hide survey when URL no longer matches
+    useEffect(() => {
+        if (isPreviewMode || !survey.conditions?.url || !posthog) {
+            return
+        }
+
+        const checkUrlMatch = () => {
+            const urlMatchType = survey.conditions?.urlMatchType ?? 'icontains'
+            const url = survey.conditions?.url
+            if (!url) {
+                return
+            }
+            const urlCheck = surveyUrlValidationMap[urlMatchType](url)
+            if (!urlCheck) {
+                setIsPopupVisible(false)
+                removeSurveyFromFocus(survey.id)
+            }
+        }
+
+        // Listen for browser back/forward and hash changes (for hash-based routing)
+        window.addEventListener('popstate', checkUrlMatch)
+        window.addEventListener('hashchange', checkUrlMatch)
+
+        // Listen for SPA navigation
+        const originalPushState = window.history.pushState
+        const originalReplaceState = window.history.replaceState
+
+        window.history.pushState = function (...args) {
+            originalPushState.apply(this, args)
+            checkUrlMatch()
+        }
+
+        window.history.replaceState = function (...args) {
+            originalReplaceState.apply(this, args)
+            checkUrlMatch()
+        }
```

Comment:
since we also have the popstate and hashchange events, does that mean we're doing the check `checkUrlMatch` twice in some cases? or its either the former (popstate, hashchange) or only pushState/replaceState?

Reply:
no, they're mutually exclusive. `hashchange` events are meant for certain SPA frameworks that use hash-based routing, but they don't fire both at the same time.